### PR TITLE
get groups from the KnowBe4 API

### PR DESCRIPTION
**Pull request by briskt**

### CODE REVIEW

File: archiver/main.go [hunk 1]
```diff
@@ -265,6 +269,52 @@ func getAllCampaigns(config LambdaConfig) ([]KnowBe4FlatCampaign, error) {
...
+		}
+
+		allGroups = append(allGroups, c...)
+
+		if len(c) < countPerPage {
+			break
+		}
+	}
+
+	flatResults, err := flattenGroups(allGroups)
```
Comment (Baggerone) [inline archiver/main.go:315]: I suppose you could just `return flattenGroups` unless you want to modify the error message.

File: archiver/main.go [hunk 2]
```diff
@@ -23,20 +23,24 @@ const (
 	countPerPage     = 500
 	maxErrorsAllowed = 5
 
+	// https://developer.knowbe4.com/reporting/#tag/Groups/paths/~1v1~1groups/get
```
Comment (jason-jackson) [inline archiver/main.go]: Do we want them listed alphabetically?

File: archiver/main.go [hunk 3]
```diff
@@ -497,9 +559,9 @@ func flattenRecipient(recipient KnowBe4Recipient) (KnowBe4FlatRecipient, error)
 	return flatRecipient, nil
 }
 
-func flattenCampaigns(recipients []KnowBe4Campaign) ([]KnowBe4FlatCampaign, error) {
-	flatCampaigns := make([]KnowBe4FlatCampaign, len(recipients))
-	for i, recipient := range recipients {
+func flattenCampaigns(campaigns []KnowBe4Campaign) ([]KnowBe4FlatCampaign, error) {
+	flatCampaigns := make([]KnowBe4FlatCampaign, len(campaigns))
+	for i, recipient := range campaigns {
```
Comment (Baggerone) [inline archiver/main.go]: should `recipient` here be `campaign`  or just `c`?

**Review by jason-jackson — APPROVED** [review]

**Review by Baggerone — APPROVED** [review]

File: archiver/main.go (hunk 2 above)
Comment (briskt) [reply archiver/main.go]: I tried, but I guess I forgot my ABCs

File: archiver/main.go [hunk 4]
```diff
@@ -265,6 +269,52 @@ func getAllCampaigns(config LambdaConfig) ([]KnowBe4FlatCampaign, error) {
...
+
+	return groups, nil
+}
+
+func getAllGroups(config LambdaConfig) ([]KnowBe4FlatGroup, error) {
+	var allGroups []KnowBe4Group
+
+	for i := 1; ; i++ {
+		c, err := getGroupsPage(i, config)
+		if err != nil {
```
Comment (jason-jackson) [inline archiver/main.go:303]: I'm curious what happens if a page is full, but there aren't any more pages afterwards, will this throw an error?

File: archiver/main.go (hunk 1 above)
Comment (briskt) [reply archiver/main.go:315]: I'm about to change this anyway, but good call.

File: archiver/main.go (hunk 4 above)
Comment (briskt) [reply archiver/main.go:303]: Good question. I just tried with Insomnia, giving a page number beyond the end of the range of data and it responded with 200 and an empty list.

File: archiver/main.go (hunk 3 above)
Comment (briskt) [reply archiver/main.go]: Yep

**Pull request merged**